### PR TITLE
Replace legacy API implementation with files()

### PR DESCRIPTION
**Pull request by FFY00**

### CODE REVIEW

**Comment by FFY00** [comment]
@jaraco the first commit is bad, but I wonder if is something that we actually want to do? The `Path` class is clearly designed to not be used, but it needs to be working because the `files()` will use it.

1) Replace `DegenerateFiles.Path` with `pathlib.Path` and force `ResourceReader.contents()` to iterate valid paths.

2) Make `DegenerateFiles.Path` try to emulate the normal behavior with non-existent `ResourceReader.contents()` files.

I would greatly prefeer 1), which actually seems like the correct behavior, if I understand `wrap_spec` correctly. 2) is a bit tricky to implement, and I am having some issues.

**Comment by jaraco** [comment]
@FFY00: Can we back up a bit? What expectation is missed that led you to draft the first commit (193ede98aa537dcf9220a7692b154f3bd2556df6)? I see the commit message says "fix DegenerateFiles not exposing loader contents", but I'm unsure why that's needed. Would it be possible to simply deprecate `contents` altogether?

**Comment by FFY00** [comment]
Currently, tests are failing if I switch the `contents` implementation to the `files` API.

https://github.com/python/importlib_resources/blob/9822338e8de57e99aa3c26275ebdbf8925af7bc8/importlib_resources/tests/test_resource.py#L56

The current `contents` implementation will work fine if we create a module manually, as done in `util.create_package`, the `files` API will not. It will return an empty iterable because that is what the `DegenerateFiles.Path` implementation does.

https://github.com/python/importlib_resources/blob/e3b5f8d8ea86c8e14edfe98129c96532e54a5a25/importlib_resources/tests/util.py#L54

We could simply deprecate `contents`, but as discussed in #80, I am reimplementing the legacy API to make sure there is no edge case where it doesn't, and as discovered here, there is.
I guess the question now is, do we care about this case? If we do, do we handle it in the `file` API, or do we come up with other solution?
To me, it seems wrong that `DegenerateFiles` will return a traversable that does not correctly represent the spec contents.

**Comment by jaraco** [comment]
If `DegenerateFiles.Path` implements this behavior, then it betrays its namesake. That is, it's no longer degenerate if it's passing through to some underlying implementation (i.e. `ResourceReader.contents`).

Maybe the right thing to do here is to rename `DegenerateFiles` to `CompatibilityFiles` or `FallbackFiles` and then allow it to grow this functionality.

An important thing to remember regarding the `ResourceReader` is that it's an abstract concept. The tests here are capturing an abstract concept that other loaders _may_ be implementing. So option (1) isn't viable (you can't assume the contents of a module could be represented by a `pathlib.Path` object).

As I see it, the question is - what should happen for a loader that implements 'contents' but not 'files' - should we wrap that implementation and create a `files`-like interface?  My instinct is no... but that will break compatibility, and it's not even be possible in the general case (`files` implements features that the other APIs were incapable of supporting).

I think you're struggling with a lot of the challenges I've struggled with while trying to adapt everything to run on `files()`.

I don't have a good recommendation at this time, but I'll try to loop back on it again later... and I welcome you to continue to explore it.

**Comment by FFY00** [comment]
> Maybe the right thing to do here is to rename `DegenerateFiles` to `CompatibilityFiles` or `FallbackFiles` and then allow it to grow this functionality.

I agree.

> As I see it, the question is - what should happen for a loader that implements 'contents' but not 'files' - should we wrap that implementation and create a `files`-like interface? My instinct is no... but that will break compatibility, and it's not even be possible in the general case (`files` implements features that the other APIs were incapable of supporting).

The thing is, this is already supported by the legacy API. I gave it another try and the issue I am running into is a lack of means to implement `is_file`, `exists`, `is_dir` -- they would need to be supported by the resource reader. Which is essentially the issue you described.

We could support these cases in the `files` API by making the returned traversable raise an unsupported operation error on such scenarios, which I feel is not a particularly bad idea.
The legacy `contents` API would return strings, so we could return traversables that would still hold that string value but have limited value besides that, we could still support `open` as `ResourceReader` can do it, but not `is_file` and friends.

This proposal makes it possible to easily replace `contents` usages with `files`, the possible issue here would be this somehow damaging expectations of the `files` API. But IMO this should not really be a big issue, as it is an edge case, and we could possibly add support to `ResourceReader` to optionally implement the missing functionality.
What do you think?

**Comment by FFY00** [comment]
I have a successful implementation that uses the approach I described.

**Comment by jaraco** [comment]
> we could possibly add support to `ResourceReader` to optionally implement the missing functionality

Since other resource providers implement the the concrete `ResourceReader` subclasses, any change would need to be compatible with the existing interface, so I suspect this approach is not worthwhile.

I notice the diffcov tests are failing with:

> importlib_resources/_adapters.py (52.2%): Missing lines 28,40,56,61-62,65-66,69,76-77,80

Not a hard requirement, but I'd prefer if we had tests covering those code paths.

**Comment by FFY00** [comment]
> Since other resource providers implement the the concrete `ResourceReader` subclasses, any change would need to be compatible with the existing interface, so I suspect this approach is not worthwhile.

Yes, makes sense, we want people moving away from `ResourceReader` to `TraversableReader` so extending `ResourceReader` does not make that much sense.

Discussion moved to #226.

**Comment by FFY00** [comment]
This is still missing tests for `CompatibilityFiles`, but is ready for review.

**Comment by FFY00** [comment]
@jaraco friendly ping. Please have a look when you have time, I know it's a pretty big, and involved, change so don't feel pressured :sweat_smile:

**Comment by FFY00** [comment]
@jaraco friendly ping.

**Comment by jaraco** [comment]
I haven’t forgotten, but I’ve had quite a lot on my plate. Will get to it soon. Thanks for the patience and reminder.

File: importlib_resources/_compat.py [hunk 1]
```diff
@@ -76,7 +80,8 @@ def _file_reader(spec):
             or
             # local FileReader
             _file_reader(self.spec)
-            or _adapters.DegenerateFiles(self.spec)
+            # fallback - adapt the spec ResourceReader to TraversableReader
+            or _adapters.CompatibilityFiles(self.spec)
```
Comment (jaraco) [inline importlib_resources/_compat.py:84]: The `_compat` module is not part of the CPython implementation, so this compatibility shim will not have any effect once this is ported to CPython. As a result, a third-party loader providing `get_resource_reader` without `.files` will not stop working for the legacy API. To support this use-case, I suspect we'll want to move the `_native_reader() or CompatibilityFiles()` logic to `_common.from_package`. Maybe also `_file_reader()` (I'm trying to imagine what loader might have modules on the file system but not be using the built-in FileReader), but I'd say let's start with just `_native_reader` and `CompatibiltyFiles`. Consider implementing this compatibility shim as a `CompatibilityFiles.try_native` classmethod.

Let me know if you'd like me to take a stab at the implementation or clarify.

Also, I note that this need is actually somewhat orthogonal to this change. The implementation you have here is perfectly suitable for `importlib_resources` and just needs to be extended for `importlib.resources`, which I might decide to defer in order to get some feedback on the approach as written.

File: importlib_resources/_common.py [hunk 2]
```diff
@@ -113,3 +115,80 @@ def _(path):
     Degenerate behavior for pathlib.Path objects.
     """
     yield path
+
+
+# legacy API
```
Comment (jaraco) [inline importlib_resources/_common.py:120]: Nitpick: Consider making this a new `_legacy` module.

**Review by jaraco — APPROVED** [review]
This work is excellent. It's clean, makes some smart decisions, but most importantly sets us up to be able to deprecate the legacy API. I have only the one serious reservation, and even that isn't a blocker to move forward, so I'll defer to your preference on whether to pursue those changes in this PR or a subsequent one.

File: importlib_resources/_compat.py (hunk 1 above)
Comment (FFY00) [reply importlib_resources/_compat.py:84]: I can have a look at it later, but I think that could be done in a separate PR. We just have to coordinate if you want to sync with CPython soon.

**Comment by jaraco** [comment]
Big thanks Filipe. Releasing as v5.2.0.

File: importlib_resources/_compat.py (hunk 1 above)
Comment (jaraco) [reply importlib_resources/_compat.py:84]: > this compatibility shim will not have any effect once this is ported to CPython

On further consideration, I believe the tests already address this concern and here's how:

- [create_package](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/util.py#L63-L67) in util creates a Loader/Reader that [does not implement `.files()`](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/util.py#L17-L51).
- For that loader/reader [tests assert](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/test_compatibilty_files.py#L87-L89) that a `CompatibilityFiles` is loaded by `wrap_spec()` (as found in _adapters).
- [files() calls wrap_spec](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/_common.py#L18-L23), which [on CPython is `_adapters.wrap_spec`](https://github.com/python/importlib_resources/blob/d89e03e3800a4baf50e3f3820754046971d04ae7/Lib/importlib/_common.py#L12).

> To support this use-case, I suspect we'll want to move the `_native_reader() or CompatibilityFiles()` logic to `_common.from_package`

- The `_native_reader() or CompatibilityFiles()` [logic is present](https://github.com/python/importlib_resources/blob/d89e03e3800a4baf50e3f3820754046971d04ae7/Lib/importlib/_adapters.py#L63-L68), similar to how I was imagining:

> Consider implementing this compatibility shim as a `CompatibilityFiles.try_native` classmethod.

At this point, I no longer have any concerns about the compatibility shim not being active in the stdlib.

Comment (FFY00) [reply importlib_resources/_compat.py:84]: I was just about to comment on the bpo, yes `CompatibilityFiles._native_reader` already provides the fallback on readers with `files()`.

My concern here would maybe be the stdlib using non-`TraversableResources` readers where a more specialized reader is available (eg. `SourceFileLoader`), but I think `CompatibilityFiles` provides a perfectly adequate shim for those. If there are any issues because of this, the best solution would be to replace those instances with `TraversableResources` readers.
